docs/proposals/resource-aggregation-proxy/README.md

```diff
@@ -0,0 +1,455 @@
+    "kind": "Pod",
+    ...
+```
+
+2\. Proxy searches Pod "nginx-3" in all caches, and finds it in cluster2's cache.
+
+3\. Proxy sends request to cluster2's API server.
+
+```yaml
+PUT https://cluster2/api/v1/namespaces/default/pods/nginx-3
+{
+    "apiVersion": "v1",
+    "kind": "Pod",
+    "metadata": {
+        "name": "nginx-3",
+        "namespace": "default",
+        "resourceVersion": "5678",
+    },
+    ...
+```
+
+4\. Proxy reads response of cluster2, then sends response to client.
+
+# Test Plan
+
+<!--
+**Note:** *Not required until targeted at a release.*
+
+Consider the following in developing a test plan for this enhancement:
+- Will there be e2e and integration tests, in addition to unit tests?
+- How will it be tested in isolation vs with other components?
+
+No need to outline all of the test cases, just the general strategy. Anything
+that would count as tricky in the implementation, and anything particularly
+challenging to test, should be called out.
+
+-->
+
+# Alternatives
+
+There are two places to put proxy.
+
+(1) Put it before API server. This requires proxy itself to Authenticate.
+
+![Alternatives-before-API-server-Architecture](Alternatives-before-API-server-Architecture.svg)
+
+(2) Put it behind API server as Aggregated API server. Let API server and proxy Authenticate coopereately.
+
+![Design-Details-Architecture](Design-Details-Architecture.svg)
+
+## Comparison
+
+|                                              | before API server                                            | behind API server                                            |
+| -------------------------------------------- | ------------------------------------------------------------ | ------------------------------------------------------------ |
+| url can be the same as kubernetes api server | ⭐️ ✔                                                          | ✘                                                            |
+| supported Authentication strategies          | X509 Client Certs                                            | ⭐️ All k8s supported Authentication strategies: X509 Client Certs, Static Token, Bootstrap Token, Service Account Token, OpenID Connect Token, Webhook Token Authentication, Authenticating Proxy |
+| performance                                  | ⭐️ Higher. Much shorter api request path: client => proxy => kube-apiserver | Lower. Longer api request path: client => kube-apiserver => proxy => kube-apiserver |
+| developerment complexity                     | High. In order to support all kubernetes Authentication strategies, we need to include kubernetes Authentication module's code, and update proxy when kubernetes updates. | ⭐️ Low. Only need to write an aggregated API server.          |
+
+## Conclusion
+
+Some users may need to use Authentication strategies other than X509 Client Certs, but implement it in proxy itself will be too complex. So we consider putting the proxy behind API server a better option.
```

Comment:
> So we consider putting the proxy behind API server a better option.

Does that means some changes are needed from `PaaS components`?

Reply:
The only changes for `PaaS components` is change the url of `apiserver`.
Originally, it may be
```
server=http://10.10.10.10:5443
```

Then  change it to
```
server=http://10.10.10.10:5443/apis/proxy.karmada.io/v1alpha1/globalresources/dummy/proxy
```

Reply:
em, both the API and implementation are similar to karmada-search now. 
Have you ever thought about putting this feature into karmada search?

Reply:
I have read the `karmada search` codes, it uses `Informer` to cache resource. It is difficult and complex to implement
[Semantics of resourveVersion](https://kubernetes.io/docs/reference/using-api/api-concepts/#semantics-for-get-and-list), [bookmarks](https://kubernetes.io/docs/reference/using-api/api-concepts/#watch-bookmarks), [resourceVersionMatch](https://kubernetes.io/docs/reference/using-api/api-concepts/#semantics-for-get-and-list)

In `proxy` component, i will use [k8s.io/apiserver/pkg/storage/cacher.Cacher](https://github.com/kubernetes/kubernetes/blob/340555476a4926e10c9ffac8b283cd77ca088c5d/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go#L226-L296), it is native to support all scenes of get/list/watch

Reply:
It's not my preference, just thinking about it.

For `proxy` purposes, we also use `informer` to collect resources from member clusters, right? Is it easy for karmada-search using the `k8s.io/apiserver/pkg/storage/cacher.Cacher`?
cc @liys87x here

Reply:
> For `proxy` purposes, we also use `informer` to collect resources from member clusters, right? I

No. `cacher.Cacher` also list/watch from member clusters, collecting resources.

If `karmada-search` use it, is shall start `informer` from the `cacher.Cacher`, instead of member cluster apiserver.

@liys87x please evaluate it.

Reply:
I found that `cacher.Cacher` is embedded in restOption and I am learning how it is used.

Reply:
Then I will evaluate it.

Reply:
so the karmada aggregation proxy will :
1. start informers listening to each member cluster and watch registered resources from member clusters
2. watched resources and theire changes from the member cluster will be stored into some etcd storage
3. start a cacher.Cacher with backend storage pointing to etcd storage above and serve list/watch requests

is my understanding correct?  @ikaven1024

Reply:
There is no informers and etcd in proxy. Indeed, the backend storage of `Cacher` is `kube-apiserver` in member clusters. Cacher list & watch from member apiserver, store member resources in memory.

Reply:
Store all resources from member cluster, it is a big challenge to etcd. So we don't use etcd in proxy.

Reply:
> 

thank you for your kind explanation @ikaven1024 
two more question please :  
    if cacher.Cacher is used , we will have n cacher.Cacher object for n member clusters ? 
    when a watch request is served, we will have a aggregation proxy to start watch from each cacher.Cacher object, then merge the events from each cacher.Cacher, modify resourceVersion, finally return values to the user?

Reply:
> we will have n cacher.Cacher object for n member clusters

Right



> when a watch request is served, we will have a aggregation proxy to start watch from each cacher.Cacher object, then merge the events from each cacher.Cacher, modify resourceVersion, finally return values to the user?

Right